qubes/events.py

```diff
@@ -157,50 +160,92 @@ def _fire_event_in_order(self, order, event, kwargs):
                     key=(lambda handler: hasattr(handler, 'ha_bound')),
                     reverse=True):
                 effect = func(self, event, **kwargs)
+                # "is True" to rule out unittest.mock.Mock...
+                if getattr(func, 'ha_async', False) is True:
+                    assert effect is not None
+                    async_effects.append(effect)
+                elif effect is not None:
+                    effects.extend(effect)
+        return effects, async_effects
+
+    @asyncio.coroutine
+    def _handle_async_result(self, sync_effects, async_effects):
+        ''' Helper function to handle coroutines returned by event handlers '''
+        # pylint: disable=no-self-use
+        effects = sync_effects
+        if async_effects:
+            async_tasks, _ = yield from asyncio.wait(async_effects)
+            for task in async_tasks:
+                effect = task.result()
                 if effect is not None:
                     effects.extend(effect)
         return effects
 
-    def fire_event(self, event, **kwargs):
+    def fire_event(self, event, pre_event=False, **kwargs):
         '''Call all handlers for an event.
 
-        Handlers are called for class and all parent classess, in **reversed**
+        Handlers are called for class and all parent classes, in **reversed**
+        or **true** (depending on *pre_event* parameter)
         method resolution order. For each class first are called bound handlers
         (specified in class definition), then handlers from extensions. Aside
         from above, remaining order is undefined.
 
+        This method call only synchronous handlers. If any asynchronous
+        handler is registered for the event, :py:class:``RuntimeError`` is
+        raised.
+
         .. seealso::
-            :py:meth:`fire_event_pre`
+            :py:meth:`fire_event_async`
 
-        :param str event: event identificator
+        :param str event: event identifier
+        :param pre_event: is this -pre- event? reverse handlers calling order
         :returns: list of effects
 
         All *kwargs* are passed verbatim. They are different for different
         events.
         '''
 
-        return self._fire_event_in_order(
-            itertools.chain(reversed(self.__class__.__mro__), (self,)),
-            event, kwargs)
-
-
-    def fire_event_pre(self, event, **kwargs):
-        '''Call all handlers for an event.
+        if pre_event:
+            order = itertools.chain((self,), self.__class__.__mro__)
+        else:
+            order = itertools.chain(reversed(self.__class__.__mro__), (self,))
```

Comment:
```python
order = itertools.chain((self,), self.__class__.__mro__)
if not pre_event:
  order = reversed(order)
```
Same as the last one: less places to err.

Also, I'd rename `_fire_event_in_order` to `_fire_event` and put the above logic in there, since it is shared between this and and `fire_event_async` (c.f. less places to err). And I'd inline `_handle_event_async` in the latter.